### PR TITLE
Get empty response when compress enabled and maxReceiveMessageSize be maxInt64

### PR DESCRIPTION
**Pull request by misvivek**
Problem:

#issue no: 4552
Get empty response when compress enabled and maxReceiveMessageSize be maxInt64
Action:

UT  Successful decompression
UT Decompressed size exceeds maxReceiveMessageSize
UT Error during decompression
UT Buffer overflow
make the decompress function compatible with calling function

Release note:

NA

### CODE REVIEW

File: rpc_util_test.go [hunk 1]
```diff
@@ -290,3 +293,115 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
 func BenchmarkGZIPCompressor1MiB(b *testing.B) {
 	bmCompressor(b, 1024*1024, NewGZIPCompressor())
 }
+
+type MockCompressor struct {
```
Comment (purnesh42H) [inline rpc_util_test.go]: why "M" capital? We are not going to export this struct

File: rpc_util_test.go [hunk 2]
```diff
@@ -290,3 +293,115 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+		t.Run(tt.name, func(t *testing.T) {
+			output, size, err := decompress(tt.compressor, tt.input, tt.maxReceiveMessageSize, nil)
+
+			if tt.wantedError != nil {
+				require.EqualError(t, err, tt.wantedError.Error())
+			} else {
+				require.NoError(t, err)
+			}
+
+			require.Equal(t, tt.wantedSize, size)
```
Comment (purnesh42H) [inline rpc_util_test.go]: let's only use base equality `==` instead of `require.Equal`

File: rpc_util_test.go [hunk 3]
```diff
@@ -290,3 +293,115 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+			wantedError:           errors.New("simulated io.Copy read error"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			output, size, err := decompress(tt.compressor, tt.input, tt.maxReceiveMessageSize, nil)
+
+			if tt.wantedError != nil {
+				require.EqualError(t, err, tt.wantedError.Error())
```
Comment (purnesh42H) [inline rpc_util_test.go]: same: let's use base equality

File: rpc_util_test.go [hunk 4]
```diff
@@ -290,3 +293,115 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+	return m.Size
+}
+
+func (m *MockCompressor) Name() string {
+	return "mockCompressor"
+}
+
+type ErrorReader struct{}
+
+func (e *ErrorReader) Read(p []byte) (n int, err error) {
```
Comment (purnesh42H) [inline rpc_util_test.go:327]: where/how is this `ErrorReader` being used?

File: rpc_util_test.go [hunk 5]
```diff
@@ -290,3 +293,115 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+	}{
+		{
+			name: "Successful decompression",
+			compressor: &MockCompressor{
+				DecompressedData: []byte("decompressed data"),
+				Size:             17,
+			},
+			input:                 mem.BufferSlice{}, // You can fill in the actual input here
+			maxReceiveMessageSize: 100,
+			wantedOutput: func() mem.BufferSlice {
```
Comment (purnesh42H) [inline rpc_util_test.go]: why is this a function?

File: rpc_util_test.go [hunk 6]
```diff
@@ -290,3 +293,115 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+		wantedSize            int
+		wantedError           error
+	}{
+		{
+			name: "Successful decompression",
+			compressor: &MockCompressor{
+				DecompressedData: []byte("decompressed data"),
+				Size:             17,
+			},
+			input:                 mem.BufferSlice{}, // You can fill in the actual input here
```
Comment (purnesh42H) [inline rpc_util_test.go]: what does "// You can fill in the actual input here" comment mean? Is it useful?

File: rpc_util_test.go [hunk 7]
```diff
@@ -290,3 +293,115 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+				ErrDecompress: errors.New("decompression error"),
+			},
+			input:                 mem.BufferSlice{},
+			maxReceiveMessageSize: 100,
+			wantedOutput:          nil,
+			wantedSize:            0,
+			wantedError:           errors.New("decompression error"),
+		},
+		{
+			name: "Buffer overflow",
```
Comment (purnesh42H) [inline rpc_util_test.go:368]: We need one more test case where maxReceiveSize is math.MaxInt64 but the actual size of data is way low. This is the fix we are trying to fix so you need test case for that

File: rpc_util.go [hunk 8]
```diff
@@ -877,35 +877,33 @@ func decompress(compressor encoding.Compressor, d mem.BufferSlice, maxReceiveMes
 		return nil, 0, err
 	}
 
-	// TODO: Can/should this still be preserved with the new BufferSlice API? Are
```
Comment (purnesh42H) [inline rpc_util.go:880]: don't delete this commented code. Its kept for reference

File: rpc_util_test.go [hunk 9]
```diff
@@ -290,3 +292,133 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
 func BenchmarkGZIPCompressor1MiB(b *testing.B) {
 	bmCompressor(b, 1024*1024, NewGZIPCompressor())
 }
+
+type mockCompressor struct {
```
Comment (purnesh42H) [inline rpc_util_test.go:296]: s/mockCompressor/testCompressor

File: rpc_util_test.go (hunk 4 above)
Comment (purnesh42H) [reply rpc_util_test.go:327]: @misvivek please respond to comments

@purnesh42H  
compressor: &mockCompressor{
				CustomReader: &ErrorReader{},
			},
at line 386 it used to create the compressor to cover the line of rpc_util.go to cover the out.Free()

File: rpc_util_test.go (hunk 5 above)
Comment (purnesh42H) [reply rpc_util_test.go]: @misvivek please respond to comment how you have addressed them

File: rpc_util_test.go [hunk 10]
```diff
@@ -290,3 +292,133 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+			wantedOutput: func() mem.BufferSlice {
+				smallDecompressed := []byte("small data")
+				return mem.BufferSlice{mem.NewBuffer(&smallDecompressed, nil)}
+			}(),
+			wantedSize:  1,
+			wantedError: nil,
+		},
+	}
+
+	for _, tt := range tests {
```
Comment (purnesh42H) [inline rpc_util_test.go:406]: s/tt/test

File: rpc_util_test.go [hunk 11]
```diff
@@ -290,3 +292,133 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+		t.Run(tt.name, func(t *testing.T) {
+			output, size, err := decompress(tt.compressor, tt.input, tt.maxReceiveMessageSize, nil)
+			if (err != nil) != (tt.wantedError != nil) {
+				t.Errorf("unexpected error state: got err=%v, want err=%v", err, tt.wantedError)
```
Comment (purnesh42H) [inline rpc_util_test.go]: line 411 and 413 have same error message. Why is there else?

File: rpc_util_test.go [hunk 12]
```diff
@@ -290,3 +292,133 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+			} else if err != nil && err.Error() != tt.wantedError.Error() {
+				t.Errorf("unexpected error message: got err=%v, want err=%v", err, tt.wantedError)
```
Comment (purnesh42H) [inline rpc_util_test.go]: `t.Errorf("decompress(): got err=%v, want err=%v", err, tt.wantedError)`

File: rpc_util_test.go (hunk 11 above)
Comment (purnesh42H) [reply rpc_util_test.go]: In line 411, i think what you want to do is 

```
if tt.wantedError != nil && err == nil {
     t.Errorf("decompress(): got nil, want err=%v", err, tt.wantedError)
}
```

File: rpc_util_test.go [hunk 13]
```diff
@@ -290,3 +292,133 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
...
+			}
+
+			if size != tt.wantedSize {
+				t.Errorf("decompress() size = %d, want %d", size, tt.wantedSize)
+			}
+			if len(tt.wantedOutput) != len(output) {
+				t.Errorf("decompress() output length = %d, wantendOutput %d", output, tt.wantedOutput)
```
Comment (purnesh42H) [inline rpc_util_test.go]: s/wantendOutput/want

File: rpc_util_test.go (hunk 1 above)
Comment (misvivek) [reply rpc_util_test.go]: done

File: rpc_util_test.go (hunk 2 above)
Comment (misvivek) [reply rpc_util_test.go]: changed the required.Equal to ==

File: rpc_util_test.go (hunk 6 above)
Comment (misvivek) [reply rpc_util_test.go]: removed comment

File: rpc_util_test.go (hunk 7 above)
Comment (misvivek) [reply rpc_util_test.go:368]: added  math.Maxint64  test condition

File: rpc_util.go (hunk 8 above)
Comment (misvivek) [reply rpc_util.go:880]: reverted

File: rpc_util_test.go (hunk 13 above)
Comment (misvivek) [reply rpc_util_test.go]: made this uniform like want  as wantedOutput

File: rpc_util_test.go (hunk 11 above)
Comment (misvivek) [reply rpc_util_test.go]: checking both err and tt.error are either nil or non-nil, the condition will be false now changed

File: rpc_util_test.go (hunk 12 above)
Comment (misvivek) [reply rpc_util_test.go]: changed that wantederror

File: rpc_util_test.go (hunk 10 above)
Comment (misvivek) [reply rpc_util_test.go:406]: I am not getting

**Pull request closed without merging**